src/main/proto/com/google/api/codegen/config.proto

```diff
@@ -292,6 +300,84 @@ message MethodConfigProto {
 
 }
 
+// `CodeSampleValues` defines sets of parameter values to be used in
+// generating samples.
+message CodeSampleValues {
+  // An identifier, unique to this method, for this set of values.
+  string id = 1;
+
+  // A short, user-visible name for this set of values. This name may
+  // be used in a UI menu, for example.
+  string title = 2;
+
+  // A longer description for this set of values. This name may be
+  // used in a UI hoverbox, for example.
+  string description = 3;
+
+  // A set of parameter values to be used together in samples. Each
+  // member of `parameters` can be specified in one of two formats:
+  //
+  // * `DOTTED.RESOURCE.NAME=VALUE`, to specify the entire value of
+  //   `NAME`.  If `VALUE` is not specified, it is treated as the
+  //    default zero value for `NAME`'s type.
+  // *  `DOTTED.RESOURCE.NAME#FIELD=VALUE`, if `NAME` is required to be
+  //    a specially formatted string using `FIELD` as one of possibly
+  //    several placeholders.
+  repeated string parameters = 4;
+}
+
+// `SampleConfigurations` defines combinations of "molds" (sample code
+// structures as output by the client library generator) and
+// `CodeSampleValues` that should be combined to generate samples.
+message SampleConfiguration {
+  // `Value` references one or more `CodeSampleValues` and provides a
+  // hint as to where it should be presented.
+  message Value{
+    // One or more of the `CodeSampleValues.id`s defined for this
+    // method. If more than one is specified and `explorer_sample` is
+    // set, the hint for relative order of the samples generated is
+    // taken to be the order of this list. If not specified, `ids`
+    // defaults to a list of all defined values for
+    // `MethodConfigProto.code_sample_values.id`. If any element of
+    // `ids` is invalid (i.e., is not found in
+    // `MethodConfigProto.code_sample_values.id`) and `delete` is not
+    // set, then an fatal error is raised at code-generation time.
+    repeated string ids = 1;
+
+    // Whether the samples specified by this configuration should also
+    // be used as configurable "Explorer Samples" suitable for
+    // inclusion in a UI element.
+    bool explorer_sample = 2;
+
+    // Whether the samples specified by this configuration should also
+    // be used as "Customary Samples" that are included within code
+    // code comments in the surface of each generated client library.
+    bool customary_sample = 3;
+
+    // A hint as to the user-visible position of this
+    // `SampleConfiguration` relative to other
+    // `SampleConfiguration`s. This is only used when generating
+    // "Explorer Samples", which only happens if `explorer_sample` is
+    // set.
+    uint32 position = 4;
+
+    // Whether or not to delete (rather than create) any previously
```

Comment:
Does this belong in `toolkit` or `artman`? So far, `toolkit` has had no knowledge of what's already "out there" -- that's entirely the realm of `artman`.

Reply:
Let me think about this for a bit....

Reply:
My first inclination is to leave this in the GAPIC config for the following reasons:

1. the artman config does not deal with method-level details, correct?

2. the actual location of the method samples directory is still handled by the artman config, which can move directories around with its `directory_mappings`. It is only the content of that directory that is configured here (and yes, with the new wrinkle of specifying behavior that only artman can implement if it encounters a pre-existing artifact that needs to be deleted)

3. if the GAPIC config is responsible for causing specific samples to be created, it should also be responsible for capturing the information that specific samples need to be deleted. Splitting it up seems confusing and counter-intuitive.

One alternative is to always have the generated samples directory overwrite the repository's sample directory; in that case, if you want to "delete" a sample, you simply make sure it is not configured and it will never be generated and hence will not show up in the new version of the samples directory. The problem with this is that we want to allow manual samples to be added to the samples directory, and we don't want to clobber them every time we regenerate by overwriting the samples directory entirely. If we want to go with this alternative, we have to figure out a different place for manual samples, and I fear that splitting samples for a given API+language among different directories will make the generated libraries less usable.

Thoughts?

Reply:
> the artman config does not deal with method-level details, correct?

Correct. I don't think this is doable in artman with configuration as-is.

> (and yes, with the new wrinkle of specifying behavior that only artman can implement if it encounters a pre-existing artifact that needs to be deleted)

Could you clarify this point: are you suggesting that toolkit would communicate to artman that artman needed to perform the deletion?

> if the GAPIC config is responsible for causing specific samples to be created, it should also be responsible for capturing the information that specific samples need to be deleted. Splitting it up seems confusing and counter-intuitive.

To be honest, I'm not sure I'm convinced that "delete existing sample" is something that should be captured in declarative configuration at all. One possible approach is that toolkit could generate a samples directory that could then be merged intelligently with a manual samples directory by artman. Assuming autogenerated samples are "tagged" as such, `artman` should know to delete and replace all existing autogenerated samples while leaving manual samples alone.

Another possible approach that avoids the need for configuration is to develop a lightweight CLI (integrated into `artman`) that takes an artman config as input and can handle deleting samples by ID.

Reply:
I agree with @geigerj here:

  * If we need a delete feature (we could always add this complexity later), this is not the place for it. We would be introducing a heretofore unknown assumption into toolkit (knowledge of surroundings).
  * It would be possible to `os.walk` the samples in a sample directory and delete samples that have been removed, if necessary, without touching manually written samples. This could also allow an auto-generated sample to be _modified_ and not replaced by removing the autogen comment line, which has value.

Reply:
A proposal I like better than the current design, and then replying to the specific comments you made (on the design I now favor less :wink: )


## Proposal
**tl;dr:**  (1) don't worry about preserving manual samples for the first iteration (focus on autogen); (2) reserve a filename prefix for manual samples, and toolkit will never autogen samples with that prefix; (3) to allow not-clobbering manual samples, alter artman to not overwrite files with that prefix

**Details:**

The motivation for the delete feature is to allow changing the set of autogen samples without clobbering manual samples (as a full-on directory overwrite would do) and without having a second samples directory for manual samples (which is not great usability).

So then let's use a different method to disambiguate manual from autogenerated samples: a requirement that filenames with a given prefix are always preserved by artman in the publishing process. For example, we could say that all manual sample files should begin with `ex.` and none of the autogenerated files will. Artman will take care to preserve any such files that may exist when cloning the repo.

(This could be extended to allow manual files anywhere in our repos, if we so choose, by having one or more non-cloberable-while-publishing prefixes.)

What this would mean in the short term is that:
* we do not need to implement the manual-sample-preserving functionality right away (let's call it "M1"), since this is less important than getting autogen samples out the door
  * for M1, artman can simply overwrite the whole samples directory if it chooses and not worry about preserving manual samples
* we do need to reserve the filename prefix that we will use to identify and preserve manual samples, simply so toolkit never autogens samples with that prefix, not even in M1
* when we do implement the manual-sample-preserving functionality ("M2"), artman will take care to not delete files with the given prefix

_Variant_:  Instead of a filename prefix, we could use the delimiter tags within the samples, as Luke suggests. If we do so, I maintain the tag names themselves would have to follow a naming convention for either manual or autogen samples, because we want to still allow including the tags in manual samples too (since the tags allows the sample files to include content that can be easily excluded from devsite). So I think having the prefix at the filename level makes it substantially easier for humans to see which samples are manual, and just a hair easier for artman to distinguish them (globbing filenames rather than inspecting each file, essentially).

## Replies to comments

(many of these are moot if we go with the Proposal above)

> Could you clarify this point: are you suggesting that toolkit would communicate to artman that artman needed to perform the deletion?

That had been my thinking: just as we configure samples to appear and artman takes care of copying them to the right location, we could anti-configure samples to _not_ appear, and artman would take care of deleting them if it happens to encounter them. 

> To be honest, I'm not sure I'm convinced that "delete existing sample" is something that should be captured in declarative configuration at all. 

I mostly disagree (but the Proposal avoids this issue altogether). If we use the config to declare things that should be there, we could also use it to declare things that should not be there.

> One possible approach is that toolkit could generate a samples directory that could then be merged intelligently with a manual samples directory by artman.

I'm opposed to this. Having two separate directories in the final output optimizes for our development, but not for the API client experience. Users should not have to track whether samples were done manually or automatically: they should just have a single samples location where the samples just work. And if we merge, we still have the manual samples directory just dangling somewhere with duplicate content (as far as the API client is concerned)---and we also have another location to track and manipulate.

> Assuming autogenerated samples are "tagged" as such, artman should know to delete and replace all existing autogenerated samples while leaving manual samples alone.

See Proposal.

> Another possible approach that avoids the need for configuration is to develop a lightweight CLI (integrated into artman) that takes an artman config as input and can handle deleting samples by ID.

I'm opposed to this too because, again, this would introduce method-level configuration into artman. And also because, as I mention above, where we configure the allowed samples can certainly be where we configure the disallowed (ie delete-if-you-see-it) samples.

> If we need a delete feature (we could always add this complexity later), 

I'm OK with deferring this feature until we need it (as in the Proposal)

> this is not the place for it.

I'm not yet convinced it's not.

> We would be introducing a heretofore unknown assumption into toolkit (knowledge of surroundings).

How so? The functionality of the (pre-Proposal) design is "when you go to publish: after you clone the repo, if you see any of the samples that match this delete config, remove them; then add the newly generated samples; then commit.". The delete configuration is idempotent as far as toolkit is concerned and requires no prior "knowledge" as to whether the disallowed samples exist or ever existed.  It simply says, "if you encounter it, delete it."

But again, I prefer the Proposal above :point_up:. Let me know your thoughts.